### ISSUE TITLE
OAuthPrompt.SendOAuthCardAsync fails due to endpoint error

### ISSUE THREAD

**Issue by sandrock** [microsoft/botframework-sdk#6598]
## Version

What package version of the SDK are you using.

   -   Top-level Package                                    Requested   Resolved
   - Microsoft.Bot.Builder.Dialogs                      4.19.3      4.19.3
   - Microsoft.Bot.Builder.Integration.AspNet.Core      4.19.3      4.19.3
   - Microsoft.Graph                                    5.10.0      5.10.0
   - Microsoft.Graph.Core                               3.0.6       3.0.6

## Describe the bug

The endpoint `https://api.botframework.com/api/botsignin/GetSignInResource` intermittently fails and returns a 404 with this message:

```json
{"error": {"code": "ServiceError","message": "Bot does not exist"}}
```

## To Reproduce

I created a project based on the [teams authentication sample](https://github.com/OfficeDev/Microsoft-Teams-Samples/tree/main/samples/bot-teams-authentication/csharp). The bot tries to ask the user for authentication using the `OAuthPrompt.SendOAuthCardAsync` method. The endpoint fails ~95% of times. Sources and extra info [here](https://gist.github.com/sandrock/24e0ff9c4b2e04e112a5e5ce6a8b448d).

## Expected behavior

The endpoint succeeds a few times a day. Meaning the request might be okay. I think the endpoint has a problem. 

## Screenshots

- [failure log](https://gist.github.com/sandrock/24e0ff9c4b2e04e112a5e5ce6a8b448d#file-log-failing-log)

## Additional context

I created this stackoverflow post: [BotFramework-Teams-Authentication issue `GetSignInResource` fails often](https://stackoverflow.com/q/76261538/282105) but it says that stackoverflow is more for "how-tos". Considering the problem is a bug, I think it is better to post it here.

## Tracking Status

**Dotnet SDK** [TODO]()
- [ ] PR
- [ ] Merged

**Javascript SDK** [TODO]()
- [ ] PR
- [ ] Merged

**Python SDK** [TODO]()
- [ ] PR
- [ ] Merged

**Java SDK** [TODO]()
- [ ] PR
- [ ] Merged

**Samples** [TODO]()
- [ ] PR
- [ ] Merged

**Docs** [TODO]()
- [ ] PR
- [ ] Merged

**Tools** [TODO]()
- [ ] PR
- [ ] Merged

**Comment by ramfattah** [comment]
Thanks @sandrock, I'm investigating.

**Comment by sandrock** [comment]
I created a new app (AAD + azure bot + teams manifest) again and the same issue occurs.

**Comment by ramfattah** [comment]
This appears to be services issues. 
I will notify the services team about this issue for further investigation.

Also, what is your Azure Bot Region location?

**Comment by ramfattah** [comment]
CC @boydc2014.

The `OAuthPrompt.SendOAuthCardAsync` method intermittently fails due to an endpoint error. 
Could you please assist us in investigating this issue from the services side? 

Thank you.

**Comment by sandrock** [comment]
> This appears to be services issues. I will notify the services team about this issue for further investigation.

Thanks.

> Also, what is your Azure Bot Region location?

When creating the bot, I chose "West Europe".

**Comment by ramfattah** [comment]
Hi @sandrock, 

For "West Europe" bots, there may be additional configurations for the OAuth. 
I wonder if this issue might be related to https://github.com/microsoft/botbuilder-js/issues/4418#issuecomment-1437471789

For a dotnet bot project, could you please add the following additional configurations to see if it resolves this issue:
```json

    "BotOpenIdMetadata": "https://login.botframework.com/v1/.well-known/openidconfiguration",
    "ValidateAuthority": "true",
    "ToChannelFromBotLoginUrl": "https://login.microsoftonline.com/botframework.com",
    "ToChannelFromBotOAuthScope": "https://api.botframework.com",
    "ToBotFromChannelTokenIssuer": "https://api.botframework.com",
    "OAuthApiEndpoint": "https://europe.api.botframework.com",
    "ToBotFromChannelOpenIdMetadataUrl": "https://login.botframework.com/v1/.well-known/openidconfiguration",
    "ToBotFromEmulatorOpenIdMetadataUrl": "https://login.microsoftonline.com/botframework.com/v2.0/.well-known/openid-configuration",
    "CallerId": "urn:botframework:azure"

```

Example:
![Screenshot 2023-06-27 at 10 30 36 AM](https://github.com/microsoft/botframework-sdk/assets/38049078/cc0f0c86-a73a-4a2e-8fd6-11d620f23407)

**Comment by sandrock** [comment]
Hi. I tried the configuration bit. The bot fails 100% of times with this issue:

> fail: Microsoft.AspNetCore.Server.Kestrel[13]
Connection id "0HMRSHJMS20RP", Request id "0HMRSHJMS20RP:00000002": An unhandled exception was thrown by the application.
System.ArgumentNullException: Value cannot be null. (Parameter 'uriString')
   at System.Uri..ctor(String uriString)
   at Microsoft.Bot.Connector.Authentication.UserTokenClientImpl..ctor(String appId, ServiceClientCredentials credentials, String oauthEndpoint, HttpClient httpClient, ILogger logger)
   at Microsoft.Bot.Connector.Authentication.ParameterizedBotFrameworkAuthentication.CreateUserTokenClientAsync(ClaimsIdentity claimsIdentity, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.CloudAdapterBase.ProcessActivityAsync(AuthenticateRequestResult authenticateRequestResult, Activity activity, BotCallbackHandler callback, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.CloudAdapterBase.ProcessActivityAsync(String authHeader, Activity activity, BotCallbackHandler callback, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.Integration.AspNet.Core.CloudAdapter.ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken)
   at TeamsBot.Controllers.BotController.PostAsync() in BotController.cs:line 37

Quick analysis:

- the url given to [UserTokenClientImpl](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Connector/Authentication/UserTokenClientImpl.cs#L42C7-L42C7) using parameter `oauthEndpoint`. Seems to be empty. 
- this url is set in a private field `ParameterizedBotFrameworkAuthentication._oAuthUrl`
- not sure how it gets there from the configuration file

**Comment by ramfattah** [comment]
Hi @sandrock, 

I was able to reproduce this issue and confirm this issue only occurs for Azure bots with region set to `West Europe`.


I resolved this issue by adding the following configuration to appsettings.json:
```json
  "OAuthUrl": "https://europe.api.botframework.com",
  "BotOpenIdMetadata": "https://login.botframework.com/v1/.well-known/openidconfiguration",
  "ValidateAuthority": "true",
  "ToChannelFromBotLoginUrl": "https://login.microsoftonline.com/botframework.com",
  "ToChannelFromBotOAuthScope": "https://api.botframework.com",
  "ToBotFromChannelTokenIssuer": "https://api.botframework.com",
  "ToBotFromChannelOpenIdMetadataUrl": "https://login.botframework.com/v1/.well-known/openidconfiguration",
  "ToBotFromEmulatorOpenIdMetadataUrl": "https://login.microsoftonline.com/botframework.com/v2.0/.well-known/openid-configuration",
  "CallerId": "urn:botframework:azure"
```

![image](https://github.com/microsoft/botframework-sdk/assets/38049078/3741ec53-b30f-4a35-8509-b78ecaf83e26)

Also for West Europe bots, please make sure the Azure Bot Redirect URI is set to:
https://europe.token.botframework.com/.auth/web/redirect

Example:
![image](https://github.com/microsoft/botframework-sdk/assets/38049078/b3ab4085-b9c3-4f6d-a5e7-2840ba729254)


Output:
<img src="https://github.com/microsoft/botframework-sdk/assets/38049078/13100257-f993-4adc-8a40-7fb6aaa73d72" width="400"/>

**Comment by ramfattah** [comment]
@sandrock, could you please try adding the updated configuration values to see if that address the issue?

**Comment by ramfattah** [comment]
Hi @sandrock, any status updates?

**Comment by dmvtech** [comment]
Closing due to inactivity. Please reopen if further assistance is needed.

**Comment by tobiascaliskan** [comment]
Hey together, I'm facing right now the same issue, but using TypeScript. The Azure Bot service is also deployed in West Europe. Is there a solution for JavaScript / TypeScript available?

The redirect URL is already setup to https://europe.token.botframework.com/.auth/web/redirect

**Comment by tobiascaliskan** [comment]
Hey @sandrock ,

thanks for the quick response. We are not using .Net but TypeScript for our bot. Even when using the configuration above as well as the redirect URL it is still not working. Just opened a ticket @ MS to receive some support.

In Teams, we are not receiving anything. Not even a card. The dialog status just stays at "WAITING". When trying to run it locally using the Bot Framework Emulator, we receive the login card. However, in the inspector of the Emulator, is stating now:

[15:18:07]Failed to generate an actual sign-in link: Error: Not Found
[15:18:07]Falling back to emulated OAuth token.

Hope to find the root cause of this quickly.

**Comment by andymarksonline** [comment]
Just to add to this, I have been developing a teams bot in C# based on one of the samples and it used OAuth authentication, this has been working fine for about a month.

Yesterday all of a sudden this stopped working and I'm getting 404 errors when showing the OAuthPrompt :
The OAuth prompt trying to retrieve the token:
https://api.botframework.com/api/usertoken/GetToken?userId=<obscured>&connectionName=<obscured>&channelId=msteams

And then because of the 404 error trying to get the token, then trying to get the sign in resources:
https://api.botframework.com/api/botsignin/GetSignInResource?state=<obscured>
which is also returning a 404 error.

I tried everything, but just now created a new app registration and bot from scratch in the azure portal, this time instead of "West Europe" for the bot I chose "Global".  I linked the bot up with my sample application and it works again as expected.

The next step is to create a new bot in West Europe and see if that works,

This is definitely an issue Microsfts end as the app works fine with the new bot and nothing else has changed at all.
